File: src/main/java/de/zalando/zally/rules/SuccessResponseAsJsonObjectRule.java
package de.zalando.zally.rules;

import de.zalando.zally.Violation;
import de.zalando.zally.ViolationType;
import io.swagger.models.HttpMethod;
import io.swagger.models.Operation;
import io.swagger.models.Path;
import io.swagger.models.Response;
import io.swagger.models.Swagger;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;


public class SuccessResponseAsJsonObjectRule implements Rule {
    private static final String TITLE = "Response As JSON Object";
    private static final String DESCRIPTION = "Always Return JSON Objects As Top-Level Data Structures To Support Extensibility";
    private static final ViolationType VIOLATION_TYPE = ViolationType.MUST;
    private static final String RULE_LINK = "https://zalando.github.io/restful-api-guidelines/compatibility/Compatibility.html#must-always-return-json-objects-as-toplevel-data-structures-to-support-extensibility";

    @Override
    public List<Violation> validate(Swagger swagger) {
        List<Violation> violations = new ArrayList<>();
        if (swagger != null && swagger.getPaths() != null) {
            for (Map.Entry<String, Path> path : swagger.getPaths().entrySet())
                for (Map.Entry<HttpMethod, Operation> op : path.getValue().getOperationMap().entrySet())
                    for(Map.Entry<String, Response> response : op.getValue().getResponses().entrySet())
                        try {
                            Integer httpCodeInt = Integer.valueOf(response.getKey());
                            if (httpCodeInt >= 200 && httpCodeInt < 300 && response.getValue().getSchema() != null && !"ref".equals(response.getValue().getSchema().getType())) {
                                String violationPath = path.getKey() + "/" + op.getKey() + "/" + response.getKey();
                                violations.add(new Violation(TITLE, DESCRIPTION, VIOLATION_TYPE, RULE_LINK, violationPath));
                            }
                        } catch(NumberFormatException nfe) {
                            // ignore any http codes which are not between 200 and 300
                        }

        }
        return violations;
    }

}

File: server/src/main/java/de/zalando/zally/rules/SuccessResponseAsJsonObjectRule.kt
package de.zalando.zally.rules

import de.zalando.zally.Violation
import de.zalando.zally.ViolationType
import io.swagger.models.ComposedModel
import io.swagger.models.ModelImpl
import io.swagger.models.Swagger
import io.swagger.models.properties.Property
import io.swagger.models.properties.RefProperty
import org.springframework.stereotype.Component
import java.util.ArrayList

@Component
open class SuccessResponseAsJsonObjectRule : AbstractRule() {
    private val TITLE = "Response As JSON Object"
    private val DESCRIPTION = "Always Return JSON Objects As Top-Level Data Structures To Support Extensibility"
    private val VIOLATION_TYPE = ViolationType.MUST
    private val RULE_LINK = "https://zalando.github.io/restful-api-guidelines/compatibility/Compatibility.html" +
            "#must-always-return-json-objects-as-toplevel-data-structures-to-support-extensibility"

    override fun validate(swagger: Swagger): Violation? {
        val paths = ArrayList<String>()
        for ((key, value) in swagger.paths.orEmpty()) {
            for ((method, operation) in value.operationMap) {
                for ((code, response) in operation.responses) {
                    val httpCodeInt = code.toIntOrZero()
                    if (httpCodeInt in 200..299) {
                        val schema = response.schema
                        if (schema != null && "object" != schema.type && !schema.isRefToObject(swagger)) {
                            paths.add("$key $method $code")
                        }
                    }
                }
            }
        }

        return if (paths.isNotEmpty()) Violation(this, TITLE, DESCRIPTION, VIOLATION_TYPE, RULE_LINK, paths) else null
    }

    private fun Property?.isRefToObject(swagger: Swagger) =
            if (this is RefProperty) {
                val model = swagger.definitions[simpleRef]
                (model is ModelImpl && model.type == "object") || model is ComposedModel
            } else false

    private fun String.toIntOrZero() =
            try {
                this.toInt()
            } catch (e: NumberFormatException) {
                0
            }
}
